test/cpp/test/src/rewards_contract.cpp

```diff
@@ -21,12 +21,82 @@ ServiceNodeRewardsContract rewards_contract(contract_address, provider);
 Signer signer(provider);    
 std::vector<unsigned char> seckey = utils::fromHexString(std::string(config.PRIVATE_KEY));
 //const std::string senderAddress = "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266";
-const std::string senderAddress = signer.addressFromPrivateKey(seckey);
+const std::string senderAddress = signer.secretKeyToAddressString(seckey);
 
 std::string erc20_address = utils::trimAddress(rewards_contract.designatedToken());
 ERC20Contract erc20_contract(erc20_address, provider);
 std::string snapshot_id = provider->evm_snapshot();
 
+static void resetContractToSnapshot()
+{
+    REQUIRE(provider->evm_revert(snapshot_id));
+}
+
+// Given the service node list and the state of the list derived in C++, verify
+// that the smart contract's service node list matches what we expect it to be.
+static void verifyEVMServiceNodesAgainstCPPState(const ServiceNodeList& snl)
+{
+    // NOTE: Collect SNs from smart contract
+    std::vector<ContractServiceNode>                  snInContract;
+    std::unordered_map<uint64_t, ContractServiceNode> snInContractMap;
+    {
+        snInContract.reserve(1 /*sentinel*/ + snl.nodes.size());
+        snInContract.push_back(rewards_contract.serviceNodes(0)); // Collect sentinel
+
+        for (size_t index = 0; index < snl.nodes.size(); index++) {
+            auto&    cppNode = snl.nodes[index];
+            uint64_t snID    = rewards_contract.serviceNodeIDs(cppNode.getPublicKey());
+            snInContract.push_back(rewards_contract.serviceNodes(snID));
+            snInContractMap[snID] = snInContract.back();
+        }
+    }
+
+    const ServiceNode sentinelCppNode = {};
+    REQUIRE(1 /*sentinel*/ + snl.nodes.size() == snInContract.size());
+
+    std::string const STAKING_REQUIREMENT_HEX = utils::padTo32Bytes(utils::decimalToHex(ServiceNodeRewardsContract::STAKING_REQUIREMENT));
+
+    for (size_t index = 0; index < snl.nodes.size(); index++) {
+        const ServiceNode&         cppNode = snl.nodes[index];
+        const ContractServiceNode& ethNode = snInContractMap[cppNode.service_node_id];
+
+        // NOTE: Verify the ethereum address is correct
+        {
+            std::array<unsigned char, 20> walletPKeyHex = signer.secretKeyToAddress(seckey);
+            REQUIRE(ethNode.recipient == walletPKeyHex);
+        }
+
+        // NOTE: Verify metadata
+        REQUIRE(ethNode.leaveRequestTimestamp == 0);
```

Comment:
This timestamp on the smart contract represents when the node started their 15 day unlock. There are 2 tests that actually make this leave request and do get their timestamp set. The ServiceNodeList doesnt track this on the C++ side, so unless we do some extra tracking so that it can actually be `REQUIRE(ethNode.leaveRequest == cppNode.leaveRequest)` this line probably should just disappear

Reply:
Yep, good point. I have commented it out and put a todo, we should eventually get around to mirroring it so we can test the EVM state fully.